README
Title: README of the prsnl10 template for DokuWiki
Creation date: 2010-05-26
Last Update: 2011-11-24
Author: Andreas Haerter <development [ät] andreas-haerter.com>
Web: <http://andreas-haerter.com>
     <http://andreas-haerter.com/projects/dokuwiki-template-prsnl10>
     <http://blog.andreas-haerter.com/2011/03/16/how-to-create-a-maintainable-dokuwiki-template>



PLEASE NOTE: I am NO native English speaker. If you find any errors please let
             me know (e.g. by mail). Otherwise, my English is not getting better
             (as desired by myself). Thank you! -- Andreas Haerter



I. Using/Installing the template
  Using "prsnl10" is simple and works just as described at
  <http://www.dokuwiki.org/template>: extract the downloaded archive into
  your DokuWiki's "lib/tpl/" folder. Then select the template in the Config
  Manager by adjusting the template option. Done. :-)



II. Customizing the template
  In order to make updates of your installed "prsnl10" template easy, follow
  these hints for customizing the template:

  - Place your files in "prsnl10/user"
    The place for *all* userdefined, custom stuff is "prsnl10/user".
    Reason: this folder will not be touched on updates. Therefore you may delete
    everything except "prsnl10/user" and copy all new files into "prsnl10/" to
    update without loosing your changes. Do not forget to empty your browser's
    cache afterwards.

  - Own logo
    If you want to replace the default text showed in the headline, simply
    create a "logo.[png|gif|jpg]" in "prsnl10/user". The template recognizes
    the file automatically and will use it instead of the default text.
    Do not forget to empty your browser's cache afterwards.
    HINT: If you want to use a text-based logo, you might find the font files
          for "Droid Sans" useful. You can download them at
          <http://www.google.com/webfonts/family?family=Droid+Sans>. Have a look
          at <http://j.mp/inYDGL> if you want to know more about.

  - Own favicon
    If you want to replace the default favicon, simply create a "favicon.ico" in
    "prsnl10/user". The template recognizes the file automatically and will use
    it instead of the default one. Do not forget to empty your browser's cache
    afterwards. By the way: you might find the following online tool useful:
    <http://tools.dynamicdrive.com/favicon/>
    Note: "favicon.png" is also possible but not recommended (cause it is not
          supported by most MSIE versions).

  - Own CSS
    To apply additional CSS rules, edit "prsnl10/user/screen.css" (influences
    normal layout) and/or "prsnl10/user/print.css" (influences print layout)
    and/or "prsnl10/user/rtl.css" (influences "right-to-left" languages like
    hebrew). These files will be included automatically (as the last ones for
    the relevant media, therefore you are able to overrule existing styles). Do
    not forget to empty your browser's cache afterwards.

  - Own JavaScript
    To load additional JavaScript, edit "prsnl10/user/user.js". Do not forget
    to empty your browser's cache afterwards.
    ATTENTION: Do not forget to activate the template option
               "prsnl10_loaduserjs" (->"Load 'prsnl10/user/user.js'?") in the
               DokuWiki Config Manager!

  - Web analytics software (Piwik, Google Analytics etc.)
    Tools like Piwik and Google Analytics normally provide a small code snippet
    ("tag") which you have to include at your website to make everything work.
    If you want to use such software with this template, have a look at
    "prsnl10/user/tracker.php". This file exists to make it easier for you.
    You don't have to deliberate where you have to copy and paste the code,
    simply do it there.

  - Own Apple Touch Icon
    If you want to replace the default Apple Touch Icon, simply create a
    "apple-touch-icon.png" in "prsnl10/user". The template recognizes the file
    automatically and will use it instead of the default one. Have a look at
    Apple's developer database if you need more information ("Configuring Web
    Applications" <http://j.mp/sx3NMT>).



III. Some notes about the dirs of this template
  * "prsnl10/conf"
    Contains configuration files:
    default.php:  default settings, see
                  <http://www.dokuwiki.org/devel:configuration#default_settings>
    metadata.php: configuration metadata, see
                  <http://www.dokuwiki.org/devel:configuration#configuration_metadata>
  * "prsnl10/lang"
    Language files (subdir names = language code). You may copy the English
    language files and translate them if your language is missing. Simply mail
    me (Andreas Haerter) the translated files and I will include them in the
    next release :-D.
  * "prsnl10/css"
    All Cascading Style Sheets (CSS). See <http://j.mp/eq8zSo> for more
    information about how the CSS is handled.
  * "prsnl10/images"
    Images files.
  * "prsnl10/js"
    JavaScript files.
  * "prsnl10/user"
    Place for userdefined stuff. E.g. if you want use an own logo, simply copy a
    logo.[png|gif|jpg] in here. See "II. Customizing the template" for details.



IV. Other notes
  - Even if it is not required: If you like the template, think about a small
    donation (see <http://andreas-haerter.com/donate/dokuwiki-template-prsnl10>)
    and/or a link to <http://andreas-haerter.com/projects/dokuwiki-template-prsnl10>.
    Thanks :-).

  - This template follows the DokuWiki coding style (see
    <http://www.dokuwiki.org/devel:coding_style>). If you want to submit any
    patches, please respect its rules (even if you don't like it - I also don't
    love it but this simply does not matter).

  - Feel free to send patches or suggestions via mail (or look out for report
    possibilities on <http://www.dokuwiki.org/template:prsnl10>).

  - Please don't delete the "index.html" and ".htaccess" files for security
    reasons, even if you don't know why they are there. These files should
    prevent others from listing directory contents (a built in feature of most
    webservers) and providing additional access protection for some files. Even
    if everything should be secure without, I prefer some additional
    precautions.



V. About
  "prsnl10" is a minimalist template for DokuWiki. It is designed to use
  DokuWiki as small CMS for portfolio websites and for other "non-wiki"
  purposes. It was created by me (Andreas Haerter), mainly for my own website at
  <http://andreas-haerter.com>. The name means "personal template, year 2010" 
  If you have any suggestions and/or bug reports, look out for report
  possibilities on <http://www.dokuwiki.org/template:prsnl10>, feel free to send
  me an email or have a look at <http://andreas-haerter.com/contact>.



VI. Creation transcript
  Some developers may find this interesting. Additionally, I am very oblivious
  :-P.

  - In principle, I followed <http://j.mp/eq8zSo>. There was just no "starter"
    template when "prsnl10" was created. Therefore, after ~1 year, I switched
    the CSS base from DokuWiki's default template to
    <http://www.dokuwiki.org/template:starter> but the markup of "main.php" is
    still based the "main.php" out of the "default" template.

  - Worked worked worked.... and done! :-P



ChangeLog
Title: Development changelog of the prsnl10 template for DokuWiki
Creation Date: 2010-05-26
Last Update: 2011-11-25
Author: Andreas Haerter <development [ät] andreas-haerter.com>
Web: <http://andreas-haerter.com>
     <http://andreas-haerter.com/projects/dokuwiki-template-prsnl10>



ATTENTION: I will not update this file. Have a look at
           <https://github.com/andreashaerter/dokuwiki-template-prsnl10/> and
           <http://www.dokuwiki.org/template:prsnl10#version_history> if you
           want to know what changed under the hood.



main.php
<?php

/**
 * Main file of the "prsnl10" template for DokuWiki
 *
 *
 * LICENSE: This file is open source software (OSS) and may be copied under
 *          certain conditions. See COPYING file for details or try to contact
 *          the author(s) of this file in doubt.
 *
 * @license GPLv2 (http://www.gnu.org/licenses/gpl2.html)
 * @author Andreas Haerter <development@andreas-haerter.com>
 * @link http://andreas-haerter.com/projects/dokuwiki-template-prsnl10
 * @link http://www.dokuwiki.org/template:prsnl10
 * @link http://www.dokuwiki.org/devel:templates
 * @link http://www.dokuwiki.org/devel:coding_style
 * @link http://www.dokuwiki.org/devel:environment
 * @link http://www.dokuwiki.org/devel:action_modes
 * @link http://blog.andreas-haerter.com/2011/03/16/how-to-create-a-maintainable-dokuwiki-template
 */


//check if we are running within the DokuWiki environment
if (!defined("DOKU_INC")){
    die();
}


/**
 * Stores the name the current client used to login
 *
 * @var string
 * @author Andreas Haerter <development@andreas-haerter.com>
 */
$loginname = "";
if (!empty($conf["useacl"])){
    if (isset($_SERVER["REMOTE_USER"]) && //no empty() but isset(): "0" may be a valid username...
        $_SERVER["REMOTE_USER"] !== ""){
        $loginname = $_SERVER["REMOTE_USER"]; //$INFO["client"] would not work here (-> e.g. when
                                              //current IP differs from the one used to login)
    }
}


//get needed language array
include DOKU_TPLINC."lang/en/lang.php";
//overwrite English language values with available translations
if (!empty($conf["lang"]) &&
    $conf["lang"] != "en" &&
    file_exists(DOKU_TPLINC."/lang/".$conf["lang"]."/lang.php")){
    //get language file (partially translated language files are no problem
    //cause non translated stuff is still existing as English array value)
    include DOKU_TPLINC."/lang/".$conf["lang"]."/lang.php";
}


//detect revision
$rev = (int)$INFO["rev"]; //$INFO comes from the DokuWiki core
if ($rev < 1){
    $rev = (int)$INFO["lastmod"];
}

?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="<?php echo hsc($conf["lang"]); ?>" lang="<?php echo hsc($conf["lang"]); ?>" dir="<?php echo hsc($lang["direction"]); ?>">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<title><?php tpl_pagetitle(); echo " - ".hsc($conf["title"]); ?></title>
<link rel="stylesheet" media="all" type="text/css" href="<?php echo ((!empty($_SERVER["HTTPS"]) && $_SERVER["HTTPS"] === "on") ? "https" : "http"); ?>://fonts.googleapis.com/css?family=Droid+Sans" />
<?php
//show meta-tags
tpl_metaheaders();
echo "<meta name=\"viewport\" content=\"width=device-width,initial-scale=1\" />";

//include default or userdefined favicon
//
//note: since 2011-04-22 "Rincewind RC1", there is a core function named
//      "tpl_getFavicon()". But its functionality is not really fitting the
//      behaviour of this template, therefore I don't use it here exclusively.
if (file_exists(DOKU_TPLINC."user/favicon.ico")){
    //user defined - you might find http://tools.dynamicdrive.com/favicon/
    //useful to generate one
    echo "\n<link rel=\"shortcut icon\" href=\"".DOKU_TPL."user/favicon.ico\" />\n";
}elseif (file_exists(DOKU_TPLINC."user/favicon.png")){
    //note: I do NOT recommend PNG for favicons (cause it is not supported by
    //all browsers).
    echo "\n<link rel=\"shortcut icon\" href=\"".DOKU_TPL."user/favicon.png\" />\n";
}else{
    //default
    echo "\n<link rel=\"shortcut icon\" href=\"".(function_exists("tpl_getFavicon") ? tpl_getFavicon() : DOKU_TPL."images/favicon.ico")."\" />\n";
}

//include default or userdefined Apple Touch Icon (see <http://j.mp/sx3NMT> for
//details)
if (file_exists(DOKU_TPLINC."user/apple-touch-icon.png")){
    echo "<link rel=\"apple-touch-icon\" href=\"".DOKU_TPL."user/apple-touch-icon.png\" />\n";
}else{
    //default
    echo "<link rel=\"apple-touch-icon\" href=\"".(function_exists("tpl_getFavicon") ? tpl_getFavicon(false, "apple-touch-icon.png") : DOKU_TPL."images/apple-touch-icon.png")."\" />\n";
}

//load userdefined js?
if (tpl_getConf("prsnl10_loaduserjs")){
    echo "<script type=\"text/javascript\" charset=\"utf-8\" src=\"".DOKU_TPL."user/user.js\"></script>\n";
}
?>
<!--[if lt IE 7]><style type="text/css">img { behavior: url(<?php echo DOKU_TPL; ?>js/iepngfix/iepngfix.htc); }</style><![endif]-->
</head>

<body>
<div id="pagewrap"<?php
    if ($ACT !== "show" && //speed up: check most common action first
        ($ACT === "admin" ||
         $ACT === "conflict" ||
         $ACT === "diff" ||
         $ACT === "draft" ||
         $ACT === "edit" ||
         $ACT === "media" ||
         $ACT === "preview" ||
         $ACT === "save")){
        echo " class=\"admin\"";
    } ?>>

    <!-- start header -->
    <div id="tmpl_header">
        <div id="tmpl_header_left">
            <?php
            //include userdefined logo or show text-headline
            echo "<div id=\"tmpl_header_logo\">\n                <a href=\"".DOKU_BASE."\" name=\"dokuwiki__top\" id=\"dokuwiki__top\" accesskey=\"h\"";
            if (file_exists(DOKU_TPLINC."user/logo.png")){
                //user defined PNG
                echo "><img src=\"".DOKU_TPL."user/logo.png\" id=\"tmpl_header_logo_img\" alt=\"\"/></a>";
            }elseif (file_exists(DOKU_TPLINC."user/logo.gif")){
                //user defined GIF
                echo "><img src=\"".DOKU_TPL."user/logo.gif\" id=\"tmpl_header_logo_img\" alt=\"\"/></a>";
            }elseif (file_exists(DOKU_TPLINC."user/logo.jpg")){
                //user defined JPG
                echo "><img src=\"".DOKU_TPL."user/logo.jpg\" id=\"tmpl_header_logo_img\" alt=\"\"/></a>";
            }else{
                //default
                echo " class=\"tmpl_header_logo_txt\">".hsc($conf["title"])."</a>";
            }
            echo "\n            </div>\n";
            ?>
        </div>

        <div id="tmpl_header_right">
            <?php
            //show header navigation?
            if (tpl_getConf("prsnl10_headernav")){
                echo "<div id=\"tmpl_header_nav\">\n                ";
                //we have to show a custom navigation
                if (empty($conf["useacl"]) ||
                    auth_quickaclcheck(cleanID(tpl_getConf("prsnl10_headernav_location")))){ //current user got access?
                    //get the rendered content of the defined wiki article to use as custom navigation
                    $interim = tpl_include_page(tpl_getConf("prsnl10_headernav_location"), false);
                    if ($interim === "" ||
                        $interim === false){
                        //show creation/edit link if the defined page got no content
                        echo "[&#160;";
                        tpl_pagelink(tpl_getConf("prsnl10_headernav_location"), hsc($lang["prsnl10_fillplaceholder"]." (".tpl_getConf("prsnl10_headernav_location").")"));
                        echo "&#160;]<br />";
                    }else{
                       //show the rendered page content
                       echo $interim;
                    }
                }
                echo "\n            </div>\n";
            }
            ?>
        </div>
        <div class="clearer"></div>
    </div>
    <!-- end header -->


    <!-- start main content area -->
    <div class="dokuwiki">
        <?php html_msgarea(); ?>

        <!--[if lt IE 7]>
        <noscript>Your browser has JavaScript disabled, the page layout will be broken.
        Please enable it and reload this page or <a href="http://browser-update.org/en/update.html">update
        your browser</a>.</noscript>
        <![endif]-->

        <!-- start main content -->
        <div id="content"<?php echo (($ACT === "media") ? " class=\"mediamanagerfix\"" : ""); ?>>
            <div class="page">
            <?php
            $toc = tpl_toc(true);
            if ($toc){
                echo $toc;
                echo "<div class=\"clearer\"></div>\n";
             } ?>

<!-- start rendered page content -->
<?php
//send already created content to get a faster page rendering on the client
tpl_flush();
//show page content
tpl_content(false);
?>
<!-- end rendered page content -->
<div class="clearer"></div>


            </div>
        </div>
        <!-- end main content -->
        <div class="clearer"></div>
        <?php
        //send already created content to get a faster page rendering on the client
        tpl_flush();
        ?>

        <div id="tmpl_footer">
            <div id="tmpl_footer_actlinksleft">
                <?php
                echo "[&#160;";
                tpl_actionlink("top");
                if (actionOK("index")){ //check if action is disabled
                    echo "&#160;|&#160;";
                    tpl_actionlink("index");
                }
                echo "&#160;]";
                ?>
            </div>
            <div id="tmpl_footer_actlinksright">
                <?php
                if (!empty($loginname) ||
                    !tpl_getConf("prsnl10_hideadminlinksfromanon")){
                    echo "[&#160;";
                    tpl_actionlink("login"); //"login" handles both login/logout
                    if (!empty($INFO["writable"])){ //$INFO comes from DokuWiki core
                        echo "&#160;|&#160;";
                        tpl_actionlink("edit"); //"edit" handles edit/create/show
                    }
                    if (!empty($INFO["exists"]) &&
                        actionOK("revisions")){ //check if action is disabled
                        echo "&#160;|&#160;";
                        tpl_actionlink("revisions");
                    }
                    if (!empty($loginname) &&
                        $ACT === "show" &&
                        actionOK("subscribe")){ //check if action is disabled
                        echo "&#160;|&#160;";
                        tpl_actionlink("subscribe");
                    }
                    if ((!empty($INFO["writable"]) || //$INFO comes from DokuWiki core
                         !empty($INFO["isadmin"]) || //purpose of this template are "non-wiki" websites, therefore show this link only to users with write permission and admins
                         !empty($INFO["ismanager"])) &&
                        actionOK("media") && //check if action is disabled
                        function_exists("media_managerURL")) { //new media manager is available on DokuWiki releases newer than 2011-05-25a "Rincewind" / since 2011-11-10 "Angua" RC1
                        echo "&#160;|&#160;";
                        tpl_actionlink("media");
                    }
                    if (!empty($INFO["isadmin"]) ||  //$INFO comes from DokuWiki core
                        !empty($INFO["ismanager"])){
                        echo "&#160;|&#160;";
                        tpl_actionlink("admin");
                    }
                    if (!empty($loginname) &&
                        actionOK("profile")){ //check if action is disabled
                        echo "&#160;|&#160;";
                        tpl_actionlink("profile");
                    }
                    echo "&#160;]";
                }else{
                    echo "&#160;";
                }
                ?>
            </div>
            <div class="clearer"></div>
            <div id="tmpl_footer_metainfo">
                <!-- You are NOT allowed to remove the following prsnl10 and/or DokuWiki link/notice. Please respect this! -->
                <a href="http://andreas-haerter.com/" target="_blank">prsnl10</a> on <a href="http://www.dokuwiki.org/" target="_blank">DW</a> under the hood
                <?php
                if(!empty($INFO["exists"]) &&
                   tpl_getConf("prsnl10_showpageinfo")) {
                    echo " &#160;|&#160; ";
                    tpl_pageinfo();
                }
                if (!empty($loginname)){
                    echo " &#160;|&#160; ";
                    tpl_userinfo();
                }
                //additional footer content?
                if (tpl_getConf("prsnl10_footer")){
                    if (empty($conf["useacl"]) ||
                        auth_quickaclcheck(cleanID(tpl_getConf("prsnl10_footer_location")))){ //current user got access?
                        //get the rendered content of the defined wiki article to use as custom content
                        $interim = tpl_include_page(tpl_getConf("prsnl10_footer_location"), false);
                        if ($interim === "" ||
                            $interim === false){
                            //show creation/edit link if the defined page got no content
                            echo "<p>[&#160;";
                            tpl_pagelink(tpl_getConf("prsnl10_footer_location"), hsc($lang["prsnl10_fillplaceholder"]." (".tpl_getConf("prsnl10_footer_location").")"));
                            echo "&#160;]</p>";
                        }else{
                            //show the rendered page content
                            echo $interim;
                        }
                    }
                }
                ?>

            </div><?php
            //copyright notice
            if (tpl_getConf("prsnl10_copyright")){
                echo "\n            <div id=\"licenseinfo\">\n                ";
                if (tpl_getConf("prsnl10_copyright_default")){
                    tpl_license(false);
                }else{
                    if (empty($conf["useacl"]) ||
                        auth_quickaclcheck(cleanID(tpl_getConf("prsnl10_copyright_location")))){ //current user got access?
                        //get the rendered content of the defined wiki article to use as custom notice
                        $interim = tpl_include_page(tpl_getConf("prsnl10_copyright_location"), false);
                        if ($interim === "" ||
                            $interim === false){
                            //show creation/edit link if the defined page got no content
                            echo "[&#160;";
                            tpl_pagelink(tpl_getConf("prsnl10_copyright_location"), hsc($lang["prsnl10_fillplaceholder"]." (".tpl_getConf("prsnl10_copyright_location").")"));
                            echo "&#160;]<br />";
                        }else{
                            //show the rendered page content
                            echo $interim;
                        }
                    }
                }
                echo "\n            </div>";
            }
            ?>

        </div>
    </div>
    <!-- end main content area -->
    <div class="clearer"></div>
    <?php
    //provide DokuWiki housekeeping, required in all templates
    tpl_indexerWebBug();
    //include web analytics software
    if (file_exists(DOKU_TPLINC."/user/tracker.php")){
        include DOKU_TPLINC."/user/tracker.php";
    }
    ?>

</div>
</body>
</html>
